HighscoreWindow.cpp

```diff
@@ -0,0 +1,134 @@
+				fMessenger(messenger)
+{
+	BStringView* congratulations = new BStringView("congratulations",
+											B_TRANSLATE("Congratulations!"));
+	congratulations->SetFont(be_bold_font);
+	congratulations->SetFontSize(25);
+	congratulations->SetAlignment(B_ALIGN_HORIZONTAL_CENTER);
+	
+	BString newHighscoreText = B_TRANSLATE("New high-score: %newhighscore%");
+	newHighscoreText.ReplaceFirst("%newhighscore%", std::to_string(newHighscore).c_str());
+	BStringView* newHighScore = new BStringView("newHighScore", newHighscoreText.String());
+	newHighScore->SetFont(be_bold_font);
+	newHighScore->SetFontSize(20);
+	newHighScore->SetAlignment(B_ALIGN_HORIZONTAL_CENTER);
+	
+	BStringView* inputlabel = new BStringView("inputlabel",
+										B_TRANSLATE("Please enter your name:"));
+	inputlabel->SetAlignment(B_ALIGN_HORIZONTAL_CENTER);
+	inputlabel->SetFont(be_bold_font);
+
+	fInputBox = new
+		BTextControl("NameInput", NULL, NULL, new BMessage(H2048_NAME_ENTERED));
+	fInputBox->SetText(B_TRANSLATE("Anonymous"));
+	fInputBox->SetAlignment(B_ALIGN_HORIZONTAL_CENTER,B_ALIGN_HORIZONTAL_CENTER);
+	fInputBox->MakeFocus();
+	fInputBox->SetModificationMessage(new BMessage(H2048_CHANGE_NAME));
+	
+	BString previousWinner;
+	previousWinner << B_TRANSLATE("This is the first record!");
+	if (oldHighscorer[0]) {
+		// Replace with previous winner and score difference
+		previousWinner = B_TRANSLATE("You've beaten the previous record holder,\n"
+										"%oldHighscorer%, by %deltaScore%.");
+		previousWinner.ReplaceFirst("%oldHighscorer%", oldHighscorer);
+		previousWinner.ReplaceFirst("%deltaScore%", 
+					std::to_string(newHighscore - oldHighscore).c_str());
+	}
+
+	BStringView* previousWinnerText = new BStringView("previousWinner", previousWinner.String());
+	previousWinnerText->SetAlignment(B_ALIGN_HORIZONTAL_CENTER);	
+	
+	BButton *saveName = new BButton("savename", B_TRANSLATE("Hooray!"),
+		new BMessage(H2048_NAME_ENTERED));
+	saveName->SetFontSize(15);
+	saveName->SetFont(be_bold_font);
+	saveName->ResizeToPreferred();
+
+	BLayoutBuilder::Group<>(this, B_VERTICAL)
+		.SetInsets(B_USE_WINDOW_INSETS)
+		.AddGroup(B_VERTICAL)
+			.Add(congratulations)
+			.Add(newHighScore)
+			.Add(previousWinnerText)
+			.AddStrut(2)
+			.Add(inputlabel)
+			.Add(fInputBox)
+			.Add(saveName)
+			.AddGlue()
+			.End();
+			
+	this->ResizeToPreferred();
+}
+
+
+bool
+HighscoreWindow::QuitRequested()
+{
+	return true;
+}
+
+
+HighscoreWindow::~HighscoreWindow()
+{
+}
+
+
+void
+HighscoreWindow::MessageReceived(BMessage* message)
+{
+	switch(message->what)
+	{
+		case H2048_NAME_ENTERED:
+		{
+			// Set a default value if empty
+			if (!fInputBox->Text()[0])
+				fInputBox->SetText(B_TRANSLATE("Anonymous"));
+
+			BMessage req(H2048_NAME_REQUESTED);
+			req.AddString("playername", fInputBox->Text());
+			fMessenger.SendMessage(&req);
+			Quit();
+			break;
+		}	
+		case H2048_CHANGE_NAME:
+		{
+			BString name = fInputBox->Text();
+			if (name.CountChars() > 25)
```

Comment:
While this works, the text control losing focus means that the user cannot just backspace a bit and use a shorter user name.

I suggest this, which leaves the input carret at the end of the text field:

```
		case H2048_CHANGE_NAME:
		{
			BString name = fInputBox->Text();
			int32 count = name.CountChars();
			if (count > 25)
				fInputBox->TextView()->Delete(count - 1, count);
			break;
		}
```